### PR TITLE
Circular buffer

### PR DESCRIPTION
**Pull request by sascha2000**
Hello everbody,

please review the code. 
Please give me a feedback about the results. 

Best Regards
Sascha

### CODE REVIEW

**Comment by tralamazza** [comment]
Thank you @sascha2000 :)

File: circularBuffer.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#ifndef circularBuffer_
+#define circsularBuffer_
```
Comment (tralamazza) [inline circularBuffer.hpp]: typo

File: circularBuffer.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+#define BufferSize 5             //circular buffer size
```
Comment (tralamazza) [inline circularBuffer.hpp]: Why not a template argument?

File: circularBuffer.hpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+// Date:   28.11.2019
+// 
+// Class: Declarartion of circular buffer class. 
+//
+//**********************************************************************************
+template <class T> class circularBuffer
+{
+  public:
+
+    static T buffer[BufferSize];
```
Comment (tralamazza) [inline circularBuffer.hpp:15]: Why static fields?

File: circularBuffer.hpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    static size_t head;              //head of circular buffer
+    static size_t tail;              //tail of circular buffer
+    static bool   full;              //buffer full if true
```
Comment (tralamazza) [inline circularBuffer.hpp:18]: Why does `full` need to be a member variable?

File: makefile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+.PHONY: clean
+
+testcb: $(OBJ)
+	$(CC) $(CFLAGS) -o testcb $(OBJ) 
+
+%.o: %.cpp
+	$(CC) $(CFLAGS) -c $<
+
+clean:
+	rm -rf $(BIN) $(OBJ) testcb
```
Comment (tralamazza) [inline makefile]: What is `$(BIN)`?

File: circularBuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  if (false == full)
+  {
+    buffer[head] = item;
+    head = (head + 1) % BufferSize;
+    full = (head == tail);
+  }
+  return;
+}
+
+template <class T> T circularBuffer<T>::get()
```
Comment (tralamazza) [inline circularBuffer.cpp]: How would you change this method to not return any data when empty?

File: circularBuffer.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+template <class T> void circularBuffer<T>::put_overwrite(T item)
+{
+  buffer[head] = item;
+  if (true == full)
+  {
+    tail = (tail + 1) % BufferSize;
+  }
+  head = (head + 1) % BufferSize;
+  full = (head == tail);
+  return;
```
Comment (tralamazza) [inline circularBuffer.cpp]: What's the reason for empty `return`  statements at the end of the function?

File: circularBuffer.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#include <iostream>
+#include "circularBuffer.hpp"
+
+template <class T> T      circularBuffer<T>::buffer[BufferSize] = {};
+template <class T> size_t circularBuffer<T>::head =                0;
+template <class T> size_t circularBuffer<T>::tail =                0;
+template <class T> bool   circularBuffer<T>::full =            false;
+
+template <class T> void circularBuffer<T>::put_overwrite(T item)
```
Comment (tralamazza) [inline circularBuffer.cpp]: Why both here and in the header?

File: circularBuffer.cpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  if (true == full)
+  {
+    tail = (tail + 1) % BufferSize;
+  }
+  head = (head + 1) % BufferSize;
+  full = (head == tail);
+  return;  
+}
+
+template <class T> void circularBuffer<T>::put_discard(T item)
```
Comment (tralamazza) [inline circularBuffer.cpp]: Why both here and in the header?

File: circularBuffer.hpp (hunk 1 above)
Comment (sascha2000) [reply circularBuffer.hpp]: Sorry ....

File: circularBuffer.hpp (hunk 2 above)
Comment (sascha2000) [reply circularBuffer.hpp]: Yes could be done like:

template <class T,uint32_t BufferSize>

File: circularBuffer.hpp (hunk 3 above)
Comment (sascha2000) [reply circularBuffer.hpp:15]: You wrote "Static memory only, no malloc()/new.".

File: circularBuffer.hpp (hunk 4 above)
Comment (sascha2000) [reply circularBuffer.hpp:18]: I don't get your point. Is this the wrong place?

File: makefile (hunk 5 above)
Comment (sascha2000) [reply makefile]: Ahhh .... sorry! A copy and paste artifact .....

File: circularBuffer.cpp (hunk 7 above)
Comment (sascha2000) [reply circularBuffer.cpp]: Thats a matter of style. Some code style guides force the same. 

Anyway this is not wrong and is more alligned to the machine code than leaving it away.

File: circularBuffer.cpp (hunk 8 above)
Comment (sascha2000) [reply circularBuffer.cpp]: Sorry its a bug. circularbuffer.cpp is not needed.

File: circularBuffer.cpp (hunk 9 above)
Comment (sascha2000) [reply circularBuffer.cpp]: Sorry its a bug. circularbuffer.cpp is not needed.

File: circularBuffer.cpp (hunk 6 above)
Comment (sascha2000) [reply circularBuffer.cpp]: Could be done by returning a bool value. True if buffer was not empty, false if buffer is empty. 

static bool get(T &item);

Call by reference used for the item.

File: circularBuffer.hpp (hunk 2 above)
Comment (sascha2000) [reply circularBuffer.hpp]: Yes can be done like that. 

template <class T, uint32_t BufferSize> class circularBuffer

I must refresh my C++ knowledge.

File: circularBuffer.hpp (hunk 3 above)
Comment (sascha2000) [reply circularBuffer.hpp:15]: Cause you wrote "4. Static memory only, no malloc()/new."

File: circularBuffer.hpp (hunk 2 above)
Comment (tralamazza) [reply circularBuffer.hpp]: 👍

File: circularBuffer.hpp (hunk 4 above)
Comment (tralamazza) [reply circularBuffer.hpp:18]: The question is why do you need a `full` member variable. Can it be implemented without it?

File: circularBuffer.cpp (hunk 6 above)
Comment (tralamazza) [reply circularBuffer.cpp]: 👍

File: circularBuffer.hpp (hunk 4 above)
Comment (sascha2000) [reply circularBuffer.hpp:18]: Yes it can be implemented without, but on the cost of one entry.

File: circularBuffer.hpp (hunk 3 above)
Comment (sascha2000) [reply circularBuffer.hpp:15]: Anyway fields don't need to be static to be setup in data memory area. If they are not static but the objects are declared globally they are also stored in data memory.

File: circularBuffer.cpp (hunk 7 above)
Comment (sascha2000) [reply circularBuffer.cpp]: Should I change it?

File: circularBuffer.hpp (hunk 4 above)
Comment (sascha2000) [reply circularBuffer.hpp:18]: Should I do so?

File: circularBuffer.hpp (hunk 3 above)
Comment (sascha2000) [reply circularBuffer.hpp:15]: Should I change?

Comment (tralamazza) [reply circularBuffer.hpp:15]: No it's not necessary to change, thank you

File: circularBuffer.hpp (hunk 4 above)
Comment (tralamazza) [reply circularBuffer.hpp:18]: No it's not necessary to change, thank you

File: circularBuffer.cpp (hunk 7 above)
Comment (tralamazza) [reply circularBuffer.cpp]: No it's not necessary to change, thank you

**Comment by tralamazza** [comment]
Thanks @sascha2000, I will close this PR and continue the talk through another channel.

**Pull request closed without merging**